proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,498 @@
+  we would like to limit the allowed names.
+- The name of the workload cluster is chosen by the admin of the workload cluster (potentially a different person or
+  automated process). We don’t currently limit the string which chooses the audience name for the workload cluster, so
+  it can be any non-empty string. The Supervisor is not aware of these strings in advance.
+
+Given that, there are two kinds of potential audience confusion for the token exchange.
+
+1. The ID token issued during the original authorization flow (before token exchange) will have an audience set to the
+   name of the dynamic client. If this client’s name happened to be the same name as the name of a workload cluster,
+   then the client could potentially skip the token exchange and use the original ID token to access the workload
+   cluster via the Concierge (acting as the user who logged in). These initial ID tokens were not meant to grant access
+   to any particular cluster.
+
+   If the admin always names the dynamic clients in a consistent way which will not collide with the names of any
+   workload cluster that they also name, then this won't happen unless another admin of a workload cluster breaks the
+   naming convention. In that case, the admin of the workload cluster has invited this kind of token misuse on their
+   cluster, possibly by accident. It is unlikely that it would be by accident if the naming convention of clusters
+   included any random element, which is what the Pinniped docs recommend. This could either be solved with
+   documentation advising against these naming collisions, or by adding code to make it impossible.
+
+   We could consider inventing a way to make this initial ID token more inert. One possibility would be to take
+   advantage of the
+   [`RequiredClaims` field](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go#L117-L119)
+   of the JWT authenticator. The token exchange could be enhanced to always add a new custom claim to the JWTs that it
+   returns, such as `pinniped_allow_concierge_tcr: true`. The Concierge JWTAuthenticator could be enhanced to require
+   this claim by default, along with a configuration option to disable that requirement for users who are not using the
+   Supervisor. ID tokens returned during an initial login (authcode flow) would not include this claim, rendering them
+   unusable at the Concierge's TokenCredentialRequest endpoint. Docs could be updated to explain that users who
+   configure dynamic clients should upgrade to use the version of the Concierge which performs this new validation on
+   workload clusters, and that users who are using JWTAuthenticators for providers other than the Supervisor would need
+   to add config to disable the new validation when they upgrade.
```

Comment:
@margocrawf @enj Is there a better way to handle this case perhaps? I think this concept would work but it's sort of a breaking change for a minority of users, so maybe there's a better way. Any ideas?

Reply:
It would be neat if we could do the opposite (e.g. some claim that only exists in the ID token returned during the initial login, but not included in the exchanged tokens or in ID tokens from somewhere other than the supervisor).
Then we could check for the absence of that claim.
But it seems like that would be more work since it's not supported the way `RequiredClaims` is.